From edadc77ce64642ec18536fd5167334bc56154892 Mon Sep 17 00:00:00 2001
From: Saurabh Sengar <saurabh.singh@xilinx.com>
Date: Mon, 8 Oct 2018 10:12:47 +0530
Subject: [PATCH 0443/1852] xilinx: v4l: dma: In case of error
 xvip_pipeline_start_stop should return gracefully

commit 9a0ec6eb9df39fa0647d0a0112f599dd16337cf0 from
https://github.com/Xilinx/linux-xlnx.git

In case of any error inside the xvip_pipeline_start_stop function, it
should release the lock taken and do media graph cleanup.

Signed-off-by: Saurabh Sengar <saurabh.singh@xilinx.com>
Reviewed-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 1b7068bcf45c..5f7bf66693b4 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -118,10 +118,8 @@ static int xvip_pipeline_start_stop(struct xvip_composite_device *xdev,
 
 	/* Walk the graph to locate the subdev nodes */
 	ret = media_graph_walk_init(&graph, mdev);
-	if (ret) {
-		mutex_unlock(&mdev->graph_mutex);
-		return ret;
-	}
+	if (ret)
+		goto error;
 
 	media_graph_walk_start(&graph, entity);
 
@@ -145,11 +143,12 @@ static int xvip_pipeline_start_stop(struct xvip_composite_device *xdev,
 			if (start && ret < 0 && ret != -ENOIOCTLCMD) {
 				dev_err(xdev->dev, "s_stream is failed on subdev\n");
 				xvip_subdev_set_streaming(xdev, subdev, !start);
-				return ret;
+				goto error;
 			}
 		}
 	}
 
+error:
 	mutex_unlock(&mdev->graph_mutex);
 	media_graph_walk_cleanup(&graph);
 
-- 
2.31.1

